File: python/dwxquickfix/order.py
# -*- coding: utf-8 -*-
"""
    order.py
    @author: Darwinex Labs (www.darwinex.com), 2021-*
    
    order - A data structure to hold open orders
"""

class order():
    
    # Side: 1=buy, 2=sell
    def __init__(self, ClOrdID=None, order_type='buy_market', symbol='EURUSD', 
                 price=None, quantity=1000, min_quantity=0, deviation=0, ttl=300, 
                 openTime='', _dict=None):
    
        if _dict is None:
            self.ClOrdID = None
            if ClOrdID is not None:
                self.ClOrdID = ClOrdID
            self.openTime = openTime
            self.symbol = symbol
            
            self.price = price
            self.deviation = deviation        # maximum slippage (only applicable for limit orders)
            self.quantity = quantity
            self.leaves_quantity = quantity
            self.min_quantity = min_quantity  # set min_quantity=quantity to not allow partial fills. 
            self.order_type = order_type
            self.ttl = ttl                    # ttl: time to live, in milliseconds
            self.error = False
            self.status = None

            if order_type == 'buy_market':
                self.side = '1'
                self.type = '1'
            elif order_type == 'buy_limit':
                self.side = '1'
                self.type = '2'
            elif order_type == 'buy_stop':
                self.side = '1'
                self.type = '3'
            elif order_type == 'sell_market':
                self.side = '2'
                self.type = '1'
            elif order_type == 'sell_limit':
                self.side = '2'
                self.type = '1'
            elif order_type == 'sell_stop':
                self.side = '2'
                self.type = '1'
            else:
                self.side = '0'
                self.type = '0'
                self.error = True
                print(f'[ERROR] Order type could not be determined! order_type: {order_type}')
            if self.type != '1' and price is None:
                self.error = True
                print(f'[ERROR] Price must be given for limit and stop orders! order_type: {order_type} | price: {price}')
        else:
            self.__dict__.update(_dict)
    
    # def to_JSON(self):
        # return json.dumps(self, default=lambda o: o.__dict__)
    
    def __str__(self):
        return (f'ClOrdID: {self.ClOrdID}, symbol: {self.symbol}, order_type: {self.order_type}, '
                f'type: {self.type}, side: {self.side}, quantity: {self.quantity}, '
                f'min_quantity: {self.min_quantity}, leaves_quantity: {self.leaves_quantity}, '
                f'price: {self.price}, ttl: {self.ttl}, status: {self.status}')
        

File: java/dwxquickfix/Order.java
package dwxquickfix;

import java.util.Date;

import static dwxquickfix.Helpers.*;


public class Order {
    
    public Date openTime;
    public int ClOrdID;
    public String symbol;
    double price;
    double deviation;
    public int quantity;
    public int minQuantity;
    public String orderType;
    public int ttl;
    public boolean error;
    public boolean confirmed;
    
    public char side;
    public char type;
    
	// market orders don't need a price or a deviation. 
	public Order(String orderType, String symbol, int quantity, int minQuantity, int ttl) {
		this(0, orderType, symbol, quantity, minQuantity, ttl);
	}
	
	public Order(int ClOrdID, String orderType, String symbol, int quantity, int minQuantity, int ttl) {
		this(ClOrdID, orderType, symbol, 0, quantity, minQuantity, 0, ttl);
	}
	
	// we don't have to specify a ClOrdID. 
	public Order(String orderType, String symbol, double price, int quantity, int minQuantity, double deviation, int ttl) {
		this(0, orderType, symbol, price, quantity, minQuantity, deviation, ttl);
	}
	
	public Order(int ClOrdID, String orderType, String symbol, double price, int quantity, int minQuantity, double deviation, int ttl) {
        this.ClOrdID = ClOrdID;
        this.orderType = orderType;
        this.symbol = symbol;
        this.price = price;
        this.quantity = quantity;
        this.quantity = quantity;
        this.minQuantity = minQuantity;
        this.deviation = deviation;
        this.ttl = ttl;
        
        this.error = false;
        this.confirmed = false;
        
        if (orderType.equals("buy_market")) {
            this.side = '1';
            this.type = '1';
        } else if (orderType.equals("buy_limit")) {
            this.side = '1';
            this.type = '2';
        } else if (orderType.equals("buy_stop")) {
            this.side = '1';
            this.type = '3';
        } else if (orderType.equals("sell_market")) {
            this.side = '2';
            this.type = '1';
        } else if (orderType.equals("sell_limit")) {
            this.side = '2';
            this.type = '1';
        } else if (orderType.equals("sell_stop")) {
            this.side = '2';
            this.type = '1';
        } else {
            this.side = '0';
            this.type = '0';
            this.error = true;
            print("[ERROR] Order direction could not be determined! orderType: " + orderType);
        }
		if (this.type != '1' && price == 0) {
            this.error = true;
            print("[ERROR] Price must be given for limit and stop orders! order_type: " + type + " | price: " + price);
		}
    }
}
